### ISSUE TITLE
Ubuntu 14.04 with CoreCLR compilation problem

### ISSUE THREAD

**Issue by tjanczuk** [tjanczuk/edge#369]
@lstratman I am testing CoreCLR support on all platforms as the last step before pushing to npm. I am running into problems on Ubuntu 14.04 as described below. Do you know what may be going on?

I am using docker image `nodesource/trusty:0.12.6` which is based on Ubuntu 14.04 with node 0.12.6 pre-installed. Then I try a redux of the steps from https://github.com/tjanczuk/edge/blob/master/tools/ubuntu_12.04_clean_install.sh as follows: 

```
docker run -it --rm --name adhoc nodesource/trusty:0.12.6 bash
# from within the container:
apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 3FA7E0328081BFF6A14DA29AA6A19B38D3D831EF
echo "deb http://download.mono-project.com/repo/debian wheezy main" | tee /etc/apt/sources.list.d/mono-xamarin.list
apt-get update
apt-get -y install curl g++ pkg-config libgdiplus libunwind8 libssl-dev unzip make mono-complete git
curl -sSL https://raw.githubusercontent.com/aspnet/Home/dev/dnvminstall.sh | DNX_BRANCH=dev sh
source .dnx/dnvm/dnvm.sh
dnvm install latest -r coreclr -alias edge-coreclr
# the above installs `1.0.0-rc1-update1    coreclr x64          linux           edge-coreclr`
dnvm use edge-coreclr
CXXFLAGS="-std=c++11" npm i tjanczuk/edge
```

This installs edge and compiles for Mono (no issues) and then CoreCLR. Native bits for CoreCLR compile without issue (but require addition of the `CXXFLAGS="-std=c++11"` which is not in the script; without this setting the build fails; is C++11 really required?). The problem occurs when compiling CoreCLREmbedding: 

```
  ACTION binding_gyp_build_managed_target_compile_coreclr_embed src/CoreCLREmbedding/bin/Release/dnxcore50/CoreCLREmbedding.dll
Microsoft .NET Development Utility CoreClr-x64-1.0.0-rc1-16231


Building CoreCLREmbedding for DNXCore,Version=v5.0
  Using Project dependency CoreCLREmbedding 1.0.0
    Source: /data/node_modules/edge/src/CoreCLREmbedding/project.json

......

  Using Package dependency System.Xml.XmlDocument 4.0.0
    Source: /root/.dnx/packages/System.Xml.XmlDocument/4.0.0
    File: ref/dotnet/System.Xml.XmlDocument.dll

System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.CodeAnalysis.CSharp.Symbol.LoadAndValidateAttributes(OneOrMany`1 attributesSyntaxLists, CustomAttributesBag`1& lazyCustomAttributesBag, AttributeLocation symbolPart, Boolean earlyDecodingOnly)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.GetAttributesBag()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamedTypeSymbol.GetAttributes()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceMemberContainerTypeSymbol.ForceComplete(SourceLocation locationOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.Symbol.ForceCompleteMemberByLocation(SourceLocation locationOpt, Symbol member, CancellationToken cancellationToken)
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
   at System.Threading.Tasks.TaskReplicator.Replica.Execute()
make: *** [src/CoreCLREmbedding/bin/Release/dnxcore50/CoreCLREmbedding.dll] Error 1
make: Leaving directory `/data/node_modules/edge/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:269:23)
gyp ERR! stack     at ChildProcess.emit (events.js:110:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:1074:12)
gyp ERR! System Linux 3.13.0-32-generic
gyp ERR! command "node" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "configure" "build"
gyp ERR! cwd /data/node_modules/edge
gyp ERR! node -v v0.12.6
gyp ERR! node-gyp -v v2.0.1
gyp ERR! not ok 
edge@5.0.0-pre node_modules/edge
├── nan@2.1.0
└── edge-cs@0.3.0
```

So two questions:
1. Do you know how to fix the error?
2. Is C++ 11 dependency really required?

Thanks!

**Comment by tjanczuk** [comment]
This may be a separate issue or just my ignorance: is Mono required to install Edge on Ubuntu, even if I intend to only use CoreCLR at runtime?

**Comment by lstratman** [comment]
You definitely shouldn't need C++ 11 to compile.  I had the flag in there when I was first developing this but then ran into issues compiling on OS X with it, so I took it out and purposely refactored the couple of things I had depending on it.  I'll see if I can reproduce the compilation issue that you're seeing without that flag as well as the .NET compilation issue which I've also never seen.

Also, regarding Mono, that's not required either.  May be a good idea to offer two docker images, one with Mono and the other with CoreCLR.  Looks like the ASP.NET team has a docker image you can use as a baseline:  https://github.com/aspnet/aspnet-docker.

**Comment by tjanczuk** [comment]
More info. I attempted installing edge on Ubuntu 14.04 with just CoreCLR installed using these steps: 

```
docker run -it --rm --name adhoc nodesource/trusty:0.12.6 bash
# from within the container:
apt-get update
apt-get -y install libunwind8 unzip
# NOTE ^ without unzip, dnvminstall fails. Without libunwind8, compilation fails early on
curl -sSL https://raw.githubusercontent.com/aspnet/Home/dev/dnvminstall.sh | DNX_BRANCH=dev sh
source .dnx/dnvm/dnvm.sh
dnvm install latest -r coreclr -alias edge-coreclr
# the above installs `1.0.0-rc1-update1    coreclr x64          linux           edge-coreclr`
dnvm use edge-coreclr
CXXFLAGS="-std=c++11" npm i tjanczuk/edge
```

The above property compiles the native bits, but then fails compiling with CoreCLREmbedding with the following: 

```
Building CoreCLREmbedding for DNXCore,Version=v5.0
  Using Project dependency CoreCLREmbedding 1.0.0
    Source: /usr/src/app/node_modules/edge/src/CoreCLREmbedding/project.json

  Using Package dependency Microsoft.CSharp 4.0.1-beta-23516
    Source: /root/.dnx/packages/Microsoft.CSharp/4.0.1-beta-23516
    File: ref/dotnet5.1/Microsoft.CSharp.dll
...........

  Using Package dependency System.Xml.XmlDocument 4.0.0
    Source: /root/.dnx/packages/System.Xml.XmlDocument/4.0.0
    File: ref/dotnet/System.Xml.XmlDocument.dll

bash: line 1:  1477 Segmentation fault      (core dumped) dnu build --configuration Release
make: *** [src/CoreCLREmbedding/bin/Release/dnxcore50/CoreCLREmbedding.dll] Error 139
make: Leaving directory `/usr/src/app/node_modules/edge/build'
```

**Comment by lstratman** [comment]
OK, for starters, ignore my earlier comments on the C++ 11 flag, I was thinking of the C++ 0x flag instead.  Compilation does require the C++ 11 flag and I have updated `binding.gyp` accordingly and will send you a PR soon.

I have also tried to reproduce your docker .NET compilation error with docker running in both Windows and Linux and have been unable to do so.  I have access to an OS X box at work that I can try on Monday.  There are a few more packages that the ASP.NET team recommends that you install at <https://docs.asp.net/en/latest/getting-started/installing-on-linux.html>, but I was unable to reproduce the compilation error even when these packages were omitted.  I've gone ahead and included them in the script that I ran below just to see if it works for you:

````bash
apt-get update
apt-get -y install libunwind8 gettext libssl-dev libcurl4-openssl-dev zlib1g libicu-dev uuid-dev unzip curl
curl -sSL https://raw.githubusercontent.com/aspnet/Home/dev/dnvminstall.sh | DNX_BRANCH=dev sh && source ~/.dnx/dnvm/dnvm.sh
dnvm install latest -r coreclr
CXXFLAGS="-std=c++11" npm i tjanczuk/edge
````

Once compilation completed, I noticed an issue using Edge without Mono installed.  Even though Mono was not available, the build was still generating `edge_nativeclr.node` even though it was empty.  We were then trying to use this module by default when loading Edge instead of falling back to `edge_coreclr.node`, which would force the user to specify `EDGE_USE_CORECLR=1` when running node even if no alternative runtime was available.  The PR that I will be sending shortly will also resolve this issue:  if you don't have Mono installed, you won't need to specify `EDGE_USE_CORECLR=1`.

**Comment by lstratman** [comment]
I believe that I may have found the issue.  I tried reproducing using Docker and OS X and ran across an error similar to yours after I ran through the script above, deleted node_modules, and tried to install from medicomp/edge instead of tjanczuk/edge.  I didn't get a segfault, but I did get a .NET memory access error.  I then tried to do the same thing Docker on Windows and Linux and was able to reproduce it there as well.

I ended up trying a bunch of different things and then I noticed that the Dockerfile that the ASP.NET team provides uses Jessie (Debian) instead of Trusty (Ubuntu) as its base.  I switched to nodesource/jessie:0.12.6 when starting the container and, viola, the errors disappeared.  So it seems that the Ubuntu Docker images (or at least the one for Trusty) are not playing nice with CoreCLR.  This may be a Docker-specific thing because I've done all of my CoreCLR dev for this under Ubuntu and haven't experienced any problems like this at all.

To recap:

1. Ran `docker run -it --rm --name adhoc nodesource/jessie:0.12.6 bash`
2. Ran the following inside the container:

````bash
apt-get update
apt-get -y install libunwind8 gettext libssl-dev libcurl4-openssl-dev zlib1g libicu-dev uuid-dev unzip curl
curl -sSL https://raw.githubusercontent.com/aspnet/Home/dev/dnvminstall.sh | DNX_BRANCH=dev sh && source ~/.dnx/dnvm/dnvm.sh
dnvm install latest -r coreclr
# You can run npm i tjanczuk/edge once my PR is merged
npm i medicomp/edge
````

**Comment by tjanczuk** [comment]
@lstratman Thanks for looking into this. It is bizarre it would not compile on Ubuntu but work on Debian - I came across an MS forum answer were someone (from MS AFAIK) claimed the exact opposite was true. 

Anyhow, I am now able to install edge and everything compiles well during installation, but I hit a roadblock soon thereafter with a simple hello, world scenario:

```
root@76d7a98f5332:/# cat > test.js <<EOF
> require('edge').func('async (x) => { return "Hello, " + x.ToString(); }');
> EOF

root@76d7a98f5332:/# node test.js 

/node_modules/edge/lib/edge.js:156
    return edge.initializeClrFunc(options);
                ^
Error: Unable to resolve project from /node_modules/edge/node_modules/edge-cs/lib
    at Error (native)
    at Object.exports.func (/node_modules/edge/lib/edge.js:156:17)
    at Object.<anonymous> (/test.js:1:79)
    at Module._compile (module.js:460:26)
    at Object.Module._extensions..js (module.js:478:10)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Function.Module.runMain (module.js:501:10)
    at startup (node.js:129:16)
    at node.js:814:3
```

I believe there is one more spot that currently requires EDGE_USE_CORECLR=1 even without Mono installed: https://github.com/tjanczuk/edge-cs/blob/master/lib/edge-cs.js#L4. As a fix I think I will just force `process.env.EDGE_USE_CORECLR=1` if appropriate somewhere around here: https://github.com/tjanczuk/edge/blob/master/lib/edge.js#L3. Thoughts?

**Comment by tjanczuk** [comment]
This seems to have done it: https://github.com/tjanczuk/edge/commit/9737c98fee452784fad9d816c3da9d7ddf264f1a. 

One can now install only CoreCLR on Debian, install Edge, and run literal C# without setting any env variables.

**Comment by tjanczuk** [comment]
This now all works, will release on npm tomorrow. Thanks!
